### PR TITLE
Fix for nondet replacement on a direct return pre-remove returns

### PR DESCRIPTION
**Pull request by JohnDumbell**
As a quick summary: if remove returns hasn't been run, there was a specific situation to do with direct returns in which a nondet wasn't being detected when it should've, which this fixes up.

More interestingly, while testing this I found out there are a variety of other situations where we can't work out precisely where the nondet statement should be inserted, specifically around any situation where the value moves between variables, like this:

temp0 = CProver.nonDetInt()
temp1 = Integer.unbox(temp0)
return_variable = temp1

The above will hit an assertion because the variable we're interested in is actually temp1, while we're searching for temp0 and not finding any assignments/returns attributed to that. I haven't fixed this (and other more complicated situations) because I felt repeated band-aids for each condition as they came up wasn't the best approach. I'm not sure what _would_ be the best approach mind, but felt it was worth mentioning.

### CODE REVIEW

File: jbmc/unit/java_bytecode/java_replace_nondet/replace_nondet.cpp [hunk 1]
```diff
@@ -70,57 +87,151 @@ void validate_method_removal(
...
-      functions.function_map.at(function_name);
+    WHEN("A method assigns a local Object variable with nondetWithoutNull.")
+    {
+      load_and_test_method(
+        "replaceNondetAssignment:()V", functions, symbol_table);
+    }
 
-    goto_model_functiont model_function(
-      symbol_table, functions, function_name, goto_function);
+    // Currently causes an abort, needs to be re-enabled when logic is fixed.
```
Comment (smowton) [inline jbmc/unit/java_bytecode/java_replace_nondet/replace_nondet.cpp]: Elaborate briefly here, or reference a Github issue number.

**Review by smowton — COMMENTED** [review]
Disabled tests should reference a Github issue that describes what's wrong-- otherwise LGTM.

**Comment by JohnDumbell** [comment]
Created #2281 just to chronicle the issue I saw and added a link to the failing tests.

**Review by smowton — APPROVED** [review]

File: jbmc/src/java_bytecode/replace_java_nondet.cpp [hunk 2]
```diff
@@ -225,47 +247,60 @@ static goto_programt::targett check_and_replace_target(
...
+  else if(target_instruction->is_assign())
+  {
+    // Assume that the LHS of *this* assignment is the actual nondet variable
+    const auto &nondet_var = to_code_assign(target_instruction->code).lhs();
+
+    side_effect_expr_nondett inserted_expr(nondet_var.type());
+    inserted_expr.set_nullable(
+      instr_info.get_nullable_type() ==
+      nondet_instruction_infot::is_nullablet::TRUE);
+    target_instruction->code = code_assignt(nondet_var, inserted_expr);
```
Comment (owen-mc-diffblue) [inline jbmc/src/java_bytecode/replace_java_nondet.cpp:300]: Should you call `target_instruction->code.add_source_location()`?

**Review by owen-mc-diffblue — APPROVED** [review]

**Review by NlightNFotis — APPROVED** [review]
LGTM

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 6082ab8db7ad1f0dd9faf30cec107e2c44d0f08f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76670784
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 6082ab8db7ad1f0dd9faf30cec107e2c44d0f08f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76670784
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 8c62e850137231630ce45eea171fecea35828242).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77232454
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 91e8b89b732373c032cbb6f24ba5c55f575cae7a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77364460

**Pull request merged**